### PR TITLE
[doc] Design and verification stages documented

### LINKED ISSUE #479: [doc] Document the development stages and sign-off procedure

### PR DESCRIPTION
**Pull request by marnovandermaas**
This PR also adds checklists for D1 and V1 stages as well as a sign-off procedure.

Closes: https://github.com/lowRISC/mocha/issues/479

### CODE REVIEW

File: doc/proj/stages.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+| ROM control       | D0               | V0                     |
+| Entropy source    | D0               | V0                     |
+| KMAC              | D0               | V0                     |
+| UART              | D0               | V0                     |
+| I2C               | D0               | V0                     |
+| SPI device        | D0               | V0                     |
+| Timer             | D0               | V0                     |
+| SPI host          | D0               | V0                     |
+| PLIC              | D0               | V0                     |
+| Tag controller    | D0               | V0                     |
```
Comment (martin-velay) [inline doc/proj/stages.md:31]: I propose to reorder by alphabetical order, to find entries easily and to center the stage status.
```suggestion
| **Block name**    | **Design stage** | **Verification stage** |
|-------------------|:----------------:|:----------------------:|
| AXI crossbar      | D0               | V0                     |
| Clock manager     | D0               | V0                     |
| CVA6-CHERI        | D0               | V0                     |
| Debug module      | D0               | V0                     |
| Entropy source    | D0               | V0                     |
| GPIO              | D0               | V0                     |
| I2C               | D0               | V0                     |
| KMAC              | D0               | V0                     |
| Mailbox           | D0               | V0                     |
| PLIC              | D0               | V0                     |
| Power manager     | D0               | V0                     |
| Reset manager     | D0               | V0                     |
| ROM control       | D0               | V0                     |
| SPI device        | D0               | V0                     |
| SPI host          | D0               | V0                     |
| SRAM              | D0               | V0                     |
| Tag controller    | D0               | V0                     |
| TileLink crossbar | D0               | V0                     |
| Timer             | D0               | V0                     |
| UART              | D0               | V0                     |
```

File: doc/proj/stages.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+Moving from one stage to another requires a formal checklist and sign-off.
+The design stages are inspired by [OpenTitan's development stages](https://opentitan.org/book/doc/project_governance/development_stages.html).
+The checklists are inspired by [OpenTitan's checklists](https://opentitan.org/book/doc/project_governance/checklist/index.html).
+Slight modification to the stages and checklists were made to meet the requirements for the COSMIC project.
+
+## Current status
+
+This table shows the current design and verification stage for each block in Mocha.
+
+| **Block name**    | **Design stage** | **Verification stage** |
```
Comment (martin-velay) [inline doc/proj/stages.md:13]: Don't we need JTAG too?

File: doc/proj/stages.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+| I2C               | D0               | V0                     |
+| SPI device        | D0               | V0                     |
+| Timer             | D0               | V0                     |
+| SPI host          | D0               | V0                     |
+| PLIC              | D0               | V0                     |
+| Tag controller    | D0               | V0                     |
+
+## Sign-off procedure
+
+To advance a block from one stage to the next you must open a pull request with the checklist in a markdown file called `doc/verif/BLOCK.md`, where `BLOCK` is replaced by the block's name.
```
Comment (martin-velay) [inline doc/proj/stages.md]: I was slightly confused with the location. And I'd prefer having that into a `checklist` directory as it's not only a verification thing but a design too.
```suggestion
To advance a block from one stage to the next you must open a pull request with the checklist in a Markdown file called `mocha/doc/checklist/BLOCK.md`, where `BLOCK` is replaced by the block's name.
```

Also, what do we do with the OT checklists, shall we remove them to avoid the confusion?

File: doc/verif/stages.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+| SPI device        | D0               | V0                     |
+| Timer             | D0               | V0                     |
+| SPI host          | D0               | V0                     |
+| PLIC              | D0               | V0                     |
+| Tag controller    | D0               | V0                     |
+
+## Sign-off procedure
+
+To advance a block from one stage to the next you must open a pull request with the checklist in a markdown file called `doc/verif/BLOCK.md`, where `BLOCK` is replaced by the block's name.
+This pull request must be approved by at least three people, one of whom should be from a different organisation than the author.
```
Comment (martin-velay) [inline doc/verif/stages.md]: If I understand well, we need a cross-functional approval: SW+DD+DV?

File: doc/proj/stages.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+This pull request must be approved by at least three people, one of whom should be from a different organisation than the author.
+It should also update the table documenting the current status of each block.
```
Comment (martin-velay) [inline doc/proj/stages.md]: Add a Markdown reference to "## Current status"

File: doc/proj/stages.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+| D2  | Feature Complete | <ul> <li> All features implemented <li> Feature frozen </ul> |
+| D3  | Design Complete | <ul> <li> Lint/CDC clean, waivers reviewed <li> Design optimisation for power and/or performance complete </ul> |
+
+### D1 design sign-off checklist
+
+Checklists for signing off a block at D1.
+
+| **Item name** | **Description** |
+|---------------|-----------------|
+| SPEC_COMPLETED | Specification is 90% complete. |
```
Comment (martin-velay) [inline doc/proj/stages.md:60]: From a DV perspective, we need the spec to be ready earlier and as complete as possible. If we start to write the verification plan based on temporary/incomplete specs, we'll spend quite some time to review again and make sure we have captured everything.

So, I think this step should be part of the D0 and at D1 we only expect minor edits and clarifications.

And even from a design perspective, I strongly believe that we need specs before starting the actual coding. So the spec should be there before we move to the next D1 stage.

File: doc/proj/stages.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+| PHYSICAL_MACROS_DEFINED_80 | Physical macros for memories and analogue components are defined and roughly 80% accurate. |
+| FUNC_IMPLEMENTED | The main functional path is implemented to allow basic testing. |
+| ASSERT_KNOWN_ADDED | Assert that all outputs of the blocks are “known.” |
+| LINT_SETUP | Lint flow is set up, but it is acceptable to have warnings at this point. |
+
+*D2 and D3 checklists to be added.*
+
+## Verification stages
+
+These are the verification stages each block goes through in case a **simulation-based approach** is taken.
```
Comment (martin-velay) [inline doc/proj/stages.md]: Each block could be mixture of simulation and formal approaches. Let me try to make a proposal which works for both

File: README.md [hunk 8]
```diff
@@ -41,6 +41,7 @@ There is a dashboard of the nightly verification runs for Mocha, which can be fo
 It shows test pass rates and coverage metrics for all the blocks currently integrated.
 Initially this dashboard is expected to grow as we add more blocks and the red statuses will turn to green as we integrate more of our verification effort.
 For more information on our top-level verification framework, check out its [dedicated documentation](hw/top_chip/dv/README.md).
+The current sign-off status for each block can be found in the [design stages documentation](doc/verif/stages.md).
```
Comment (martin-velay) [inline README.md]:
```suggestion
The current sign-off status for each block can be found in the [design stages documentation](doc/checklist/stages.md).
```

File: doc/proj/stages.md (hunk 3 above)
Comment (martin-velay) [reply doc/proj/stages.md]: Or maybe we can call this directory `maturity`?

File: doc/proj/stages.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+### D1 design sign-off checklist
+
+Checklists for signing off a block at D1.
+
+| **Item name** | **Description** |
+|---------------|-----------------|
+| SPEC_COMPLETED | Specification is 90% complete. |
+| CSR_DEFINED | Registers defined for the primary programming model. |
+| CLKRST_CONNECTED | Clock and reset connected to all submodules. |
+| IP_TOP | There is an IP top that can be included in the top design. |
```
Comment (martin-velay) [inline doc/proj/stages.md:63]: Is this item related to the "ports frozen" item you've listed above?

File: doc/proj/stages.md (hunk 6 above)
Comment (martin-velay) [inline doc/proj/stages.md:60]: Another point about specs, we should probably define the specs we are expecting: system spec (if applicable), detailed design spec (micro-architectural spec), others?

File: doc/proj/stages.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+| V2 | Testing Complete | <ul> <li> Documentation: <ul> <li> Verification document completely written </ul> <li> Design issues: <ul> <li> All high priority bugs addressed <li> Low priority bugs root-caused </ul> <li> Testbench: <ul> <li> All interfaces have assertions checking the protocol <li> All functional assertions written and enabled <li> Assumptions for FPV specified and reviewed </ul> <li> Tests (written and passing): all tests planned for in the testplan <li> Regression: 90% of properties proven in nightly regression <li> Coverage: 90% code coverage and 75% logic cone of influence (COI) coverage </ul> |
+| V3 | Verification Complete | <ul> <li> Design issues: all bugs addressed <li> Assertions (written and proven): all assertions including newly added post-V2 assertions (if any) <li> Regression: 100% of properties proven (with reviewed assumptions) <li> Coverage: 100% code coverage and 100% COI coverage </ul> |
+
+### V1 verification sign-off checklist
+
+Checklist for signing off a block at V1.
+
+| **Item name** | **Description** |
+|---------------|-----------------|
+| DV_DOC_DRAFT_COMPLETED | Verification document drafted with overall goal and strategy. |
```
Comment (martin-velay) [inline doc/proj/stages.md]: We must have 2 docs:
- Verification Plan: lists **what** should be verified
- Verification Strategy: focuses on **how** to verify, detailing the tools, specific tests, stimuli and procedures to be followed to achieve the objectives set out in the verification plan...

They should be initiated from V0 and completed with reviews in V1. Note that reviewing the vPlan is a big task as it should involve all the stakeholders.

File: doc/proj/stages.md
Comment (johngt) [inline doc/proj/stages.md:1]: @marnovandermaas - on L12 you mention that a different organisation should review the sign-off.
For COSMIC we are unlikely to be able to get verification for all the blocks from other orgs. Can we caveat this with "should ideally".
In the review there should ideally be someone who has not been involved in the design and the verification of the block 'external review'

Only other comment is whether we should keep V2(S) in the table. Although security hardening is not considered we should probably mention that in the table to note a deviation from OT approach

Comment (marnovandermaas) [reply doc/proj/stages.md:1]: Yes of course, both of these make sense to me, hopefully I've addressed both of these now.

File: doc/proj/stages.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+## Design stages
+
+These are the stages each block goes through.
+
+| **Stage** | **Name** | **Definition** |
+|-----------|----------|----------------|
+| D0  | Initial Work | RTL being developed, not functional. |
+| D1  | Functional | <ul> <li> Feature set finalized, spec complete </li> <li> CSRs identified; RTL/DV/SW collateral generated </li> <li> SW interface automation completed </li> <li> Clock(s) and reset(s) connected to all sub modules </li> <li> Lint run setup </li> <li> Ports frozen </li> </ul> |
+| D2  | Feature Complete | <ul> <li> All features implemented </li> <li> Feature frozen </li> </ul> |
```
Comment (martin-velay) [inline doc/proj/stages.md:50]: A missing point: at this stage the UNR (auto-exclusion) flow should be in place and in D3 all auto-exclusions should have been reviewed. Also the manual exclusions should also be mentioned somewhere in D3 I believe

File: doc/proj/stages.md (hunk 7 above)
Comment (martin-velay) [reply doc/proj/stages.md]: It's not yet 100% correct, but most of it is there. I need to review it again later, as my brain needs some rest. WDYT in general?

## Verification stages

A block may be verified using simulation, formal-property verification (FPV), or a mix of the two.
The stages and checklist below are methodology-neutral: the verification plan declares, per vPlan item (also referred as verification point), which methodology covers it, and the checklist items are then applied accordingly.

### Verification documentation

Two documents are required for every block. Both are initiated at V0 and must be completed and reviewed for V1 sign-off.

* **Verification plan (vPlan)** - *what* must be verified.
  Lists features, and the required vPlan items to cover each feature. And also for each vPlan item, the methodology which will be used to covers it.
  Reviewing the vPlan is a substantial task and must involve all stakeholders (system, design, DV, software, security, integration), since it locks in the scope of verification for the block.
* **Verification strategy** - *how* it will be verified.
  Describes the tools, testbench / formal-environment architecture, specific tests, stimuli, assumptions, and procedures used to meet the objectives set out in the verification plan.

### Stage definitions

| **Stage** | **Name** | **Definition** |
| --- | --- | --- |
| V0 | Initial Work | <ul> <li> Verification plan and verification strategy being created <li> Methodology being decided per feature (simulation, FPV, or both) <li> Testbench and/or formal environment skeleton created (not yet functional) <li> Define DV specific requirements for the Continuous Integration (CI) pipeline </ul> |
| V1 | Initial Testing | <ul> <li> Documentation: <ul> <li> Verification Plan completed and reviewed (involves all stakeholders) <li> Verification Strategy being drafted </ul> <li> Testbench: <ul> <li> Verification environment / formal flow in place for initial testing <li> Initial functional coverage model bound in <li> VIPs / agents: main are ready and secondary are at least specified <li> Device Under Test (DUT) instantiated with major interfaces hooked up <li> Available interface assertion checkers hooked up <li> X / unknown checks on DUT outputs added </ul> <li> Tests / assertions: *Interfaces and Sanity* <ul> <li> Register / memory test suite completed <li> Sanity tests / assertions accessing the main functional paths  <li> Connectivity checks </ul> <li> Regressions set up: sanity and nightly </ul> |
| V2 | Complete Testing | <ul> <li> Documentation: <ul> <li> Verification Strategy completed and reviewed <li> Verification Plan kept up to date with any in-flight changes </ul> <li> Design issues: <ul> <li> All high-priority bugs addressed <li> Low-priority bugs root-caused, tracked as issues, and triaged for V3 </ul> <li> Environment: <ul> <li> Verification environment fully developed: scoreboard with end-to-end checks, reference / predictor models integrated <li> All interfaces exercised (if applicable: with protocol checkers enabled) <li> All FPV assumptions specified and reviewed </ul> <li> Tests / assertions: <ul> <li> ≥ 90% of vPlan items implemented (sim tests / formal assertions) and passing <li> Functional covergroups implemented for the items covered so far </ul> <li> Regression: <ul> <li> Sim: nightly regression with multiple seeds, > 90% passing <li> Formal: ≥ 90% of properties proven in nightly regression </ul> <li> Coverage: <ul> <li> Sim: ≥ 90% code coverage and ≥ 90% functional coverage; gaps documented (not yet waived) <li> Formal: ≥ 90% code coverage and ≥ 75% cone-of-influence (COI) coverage </ul> </ul> |
| V3 | Verification Complete | <ul> <li> Documentation: Verification Plan and Verification Strategy final, reviewed, and signed off <li> Design issues: <ul> <li> All bugs fixed, or formally waived with sign-off <li> No open verification-blocking issues </ul> <li> Environment: <ul> <li> No outstanding TODOs in the environment <li> All assertions enabled; any disabled assertions documented and reviewed <li> FPV assumptions finalised and reviewed </ul> <li> Tests / assertions: <ul> <li> All vPlan items, including any post-V2 additions, implemented and passing <li> Error and corner-case scenarios exercised <li> All functional covergroups defined in the vPlan implemented </ul> <li> Regression: <ul> <li> Sim: 100% pass rate on all seeds, sustained over several consecutive nightly runs <li> Formal: 100% of properties proven with reviewed assumptions </ul> <li> Coverage: <ul> <li> Sim: 100% code and 100% functional coverage, with reviewed waivers <li> Formal: 100% code and 100% COI coverage, with reviewed waivers </ul> </ul> |

### V1 verification sign-off checklist

Each item below has an **Applies to** column:

* *Both* - required regardless of methodology.
* *Sim* - required if the verification plan assigns any vPlan item to simulation.
* *Formal* - required if the verification plan assigns any vPlan item to FPV.

Items whose methodology is not used by the block are recorded as N/A. The
methodology mix declared in the V1 verification plan is locked at V1 sign-off;
changing it later requires a new vPlan review.

| **Item name** | **Applies to** | **Description** |
|---------------|----------------|-----------------|
| DV_PLAN_COMPLETED | Both | Verification plan drafted: full feature list organised into sections, with their associated coverage methodology (sim / formal / both). At this stage we don't expect to have all vPlan items defined. |
| DV_STRATEGY_COMPLETED | Both | Verification strategy drafted: tools, environment architecture, specific tests, stimuli, assumptions, and procedures used to meet the verification plan objectives. |
| DV_PLAN_REVIEWED | Both | Verification plan reviewed by all stakeholders (system, design, DV, software, security, integration). |
| DV_STRATEGY_REVIEWED | Both | Verification strategy reviewed. |
| DESIGN_SPEC_REVIEWED | Both | Design specification reviewed from a verification perspective. |
| PRE_VERIFIED_SUB_MODULES_V1 | Both | Pre-verified sub-modules must also have reached V1. |
| SIM_SMOKE_TEST_PASSING | Sim | Smoke test passing in simulation with a particular seed. |
| SIM_SMOKE_REGRESSION_SETUP | Sim | Smoke regression tests selected and defined. |
| SIM_NIGHTLY_REGRESSION_SETUP | Sim | Nightly regression tests selected and defined. |
| SIM_COVERAGE_MODEL_ADDED | Sim | Initial functional coverage model bound in. |
| FPV_MAIN_ASSERTIONS_PROVEN | Formal | Main functional-path assertions implemented and proven. Each input and each output of the module is part of at least one assertion. |
| FPV_REGRESSION_SETUP | Formal | FPV regression set up and added to `top_chip_fpv_ip_cfgs.hjson`. |

*V2 and V3 checklists to be added.*

File: doc/proj/stages.md (hunk 2 above)
Comment (marnovandermaas) [reply doc/proj/stages.md:13]: I had considered that to be part of the debug module.

File: doc/proj/stages.md (hunk 3 above)
Comment (marnovandermaas) [reply doc/proj/stages.md]: I agree it's confusing having design checklists in a verif directory. What about `proj` short for "project" which is a reference to the project governance documentation in OpenTitan?

Comment (marnovandermaas) [reply doc/proj/stages.md]: It might be useful to keep the OpenTitan checklists in tree to be able to reference them? I'm not sure whether this is useful until I actually do a sign-off though.

File: doc/verif/stages.md (hunk 4 above)
Comment (marnovandermaas) [reply doc/verif/stages.md]: Yes, this is the ideal scenario. Also it would be nice if there is someone involved from a different organisation. I think I don't want to be too prescriptive in the sign-off procedure though.

File: doc/proj/stages.md (hunk 5 above)
Comment (marnovandermaas) [reply doc/proj/stages.md]: Good idea, this is now done.

File: doc/proj/stages.md (hunk 6 above)
Comment (marnovandermaas) [reply doc/proj/stages.md:60]: There is no sign-off procedure for D0, D1 is the first stage where we can require things in a checklist. Of course the specification should come as early as possible, the checklist just enforces the latest point at which this needs to be done not the earliest.

File: README.md (hunk 8 above)
Comment (marnovandermaas) [reply README.md]: Thanks for this reminder!

File: doc/proj/stages.md (hunk 6 above)
Comment (marnovandermaas) [reply doc/proj/stages.md:60]: I think we should leave this up to the judgement of those making the sign-offs. If a micro-architectural spec is necessary then it should be included. I'm not sure what you mean by a system spec when doing a block-level sign-off.

File: doc/proj/stages.md (hunk 10 above)
Comment (marnovandermaas) [reply doc/proj/stages.md]: Although I am not opposed to having two documents in general, I am careful to make this a requirement specifically for blocks that we are importing from OpenTitan. We need to be able to re-use the sign-offs we did there otherwise we will not be able to reach our project timelines.

File: doc/proj/stages.md (hunk 9 above)
Comment (marnovandermaas) [reply doc/proj/stages.md:63]: Interesting, this was a copy from OpenTitan. I've removed the port frozen requirement because I think it is kind of unnecessary to re-do a D1 sign-off when an extra signal is added to the top.

File: doc/proj/stages.md (hunk 7 above)
Comment (marnovandermaas) [reply doc/proj/stages.md]: Good idea to add the "applies to" column. I've now added that. I've also tried to reduce the duplication between formal and the simulation tables but don't want to change the wording too much so that we can re-use OpenTitan sign-offs.

File: doc/proj/stages.md (hunk 6 above)
Comment (martin-velay) [reply doc/proj/stages.md:60]: Yes, the word system spec is confusing, it comes from my previous company doc flow.
The system spec was the high level doc where you describe the HW interface, the CSRs, the algos... this doc were derived from the requirements and made by the leads with the system team. Then based on this system spec, the design team were making their design spec (micro architectural description), and so DV team with their vPlan (note that the vPlan also contains elements from the design spec)

File: doc/proj/stages.md (hunk 11 above)
Comment (marnovandermaas) [reply doc/proj/stages.md:50]: This is not part of the OpenTitan checklist so I would prefer not to mandate it.

File: doc/proj/stages.md (hunk 1 above)
Comment (marnovandermaas) [reply doc/proj/stages.md:31]: Good idea, this should now be done.

File: doc/proj/stages.md (hunk 10 above)
Comment (martin-velay) [reply doc/proj/stages.md]: Yes, you're probably right on that point, I described more what it should be without considering project timelines. Something to discuss further though for OT too

File: doc/proj/stages.md (hunk 7 above)
Comment (marnovandermaas) [reply doc/proj/stages.md]: Ok, I've now also had a go at combining the simulation and formal tables.

**Review by martin-velay — APPROVED** [review]
I think we are good for now with this first version. Thanks Marno
